### ISSUE TITLE
Cannot load fbcdn-profile-a.akamaihd.net/hprofile++ (spdy)

### ISSUE THREAD

**Issue by anatasiajp** [urllib3/urllib3#656]
Urllib3 and urllib.request/urllib2 cannot load this url, can you implement a fix or patch?

```
import urllib3
import urllib.request

p = urllib3.PoolManager()
#urllib2 test
#r = urllib.request.urlopen("https://fbcdn-profile-a.akamaihd.net/hprofile-ak-xfp1/v/t1.0-1/p32x32/1795556_454025928112183_7252557420120250438_n.jpg?oh=2f63a1f58e29fd34605626a87fdd7202&oe=55EA9D3F&__gda__=1445088810_e91fee57938ed896996ea686b6a0e628")

r = p.urlopen('GET', "https://fbcdn-profile-a.akamaihd.net/hprofile-ak-xfp1/v/t1.0-1/p32x32/1795556_454025928112183_7252557420120250438_n.jpg?oh=2f63a1f58e29fd34605626a87fdd7202&oe=55EA9D3F&__gda__=1445088810_e91fee57938ed896996ea686b6a0e628")
print(r.data)
#urllib2 test
#print(r.read())
```

But without https the above code works. Works on Firefox.

This may interest you:

```
Request URL:    https://fbcdn-profile-a.akamaihd.net/hprofile-ak-xfp1/v/t1.0-1/p32x32/1795556_454025928112183_7252557420120250438_n.jpg?oh=2f63a1f58e29fd34605626a87fdd7202&oe=55EA9D3F&__gda__=1445088810_e91fee57938ed896996ea686b6a0e628
Request Method:     GET
Status Code:    HTTP/1.1 200 OK
Request Headers 02:35:52.000
User-Agent: Mozilla/5.0 (Windows NT 6.1; WOW64; rv:25.4) Gecko/20150509 Firefox/31.9 PaleMoon/25.4.1
Host:   fbcdn-profile-a.akamaihd.net
Connection: keep-alive
Accept-Language:    en-US,en;q=0.5
Accept-Encoding:    gzip, deflate
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Response Headers Δ350ms
X-Firefox-Spdy: 3
Timing-Allow-Origin:    *
Last-Modified:  Mon, 15 Jun 2015 15:59:16 GMT
Expires:    Wed, 08 Jul 2015 17:35:50 GMT
Date:   Wed, 24 Jun 2015 17:35:50 GMT
Content-Type:   image/jpeg
Content-Length: 1255
Cache-Control:  no-transform, max-age=1209600
Access-Control-Allow-Origin:    *
Received Image 32x32px, Δ0ms
```
##

**Comment by Lukasa** [comment]
@cattleyavns What do you mean urllib3 cannot load that URL? It loads fine for me using exactly your code. Can you elaborate on what error you encounter?

**Comment by cattleyavns** [comment]
>> What do you mean urllib3 cannot load that URL? It loads fine for me using exactly your code. Can you elaborate on what error you encounter?

No, for me it never return anything, just:

```
C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
```

And Python Interpreter never finish.

**Comment by Lukasa** [comment]
Given that this happens to the stdlib as well, I doubt this problem is directly with urllib3.

Can you set a timeout and see what happens? Try running:

```python
p = urllib3.PoolManager(timeout=5)
r = p.urlopen('GET', "https://fbcdn-profile-a.akamaihd.net/hprofile-ak-xfp1/v/t1.0-1/p32x32/1795556_454025928112183_7252557420120250438_n.jpg?oh=2f63a1f58e29fd34605626a87fdd7202&oe=55EA9D3F&__gda__=1445088810_e91fee57938ed896996ea686b6a0e628")
print(r.data)
```

**Comment by cattleyavns** [comment]
Hi, this is what I got with timeout=5:

```
>>> 
C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py:768: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
Traceback (most recent call last):
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py", line 372, in _make_request
TypeError: getresponse() got an unexpected keyword argument 'buffering'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py", line 374, in _make_request
  File "C:\Python34\lib\http\client.py", line 1171, in getresponse
    response.begin()
  File "C:\Python34\lib\http\client.py", line 351, in begin
    version, status, reason = self._read_status()
  File "C:\Python34\lib\http\client.py", line 313, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "C:\Python34\lib\socket.py", line 374, in readinto
    return self._sock.recv_into(b)
  File "C:\Python34\lib\ssl.py", line 751, in recv_into
    return self.read(nbytes, buffer)
  File "C:\Python34\lib\ssl.py", line 623, in read
    v = self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py", line 544, in urlopen
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py", line 376, in _make_request
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py", line 304, in _raise_timeout
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='fbcdn-profile-a.akamaihd.net', port=443): Read timed out. (read timeout=5)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 420, in run_nodebug
  File "<module3>", line 8, in <module>
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\poolmanager.py", line 161, in urlopen
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py", line 617, in urlopen
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py", line 617, in urlopen
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py", line 617, in urlopen
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\connectionpool.py", line 597, in urlopen
  File "C:\Python34\lib\site-packages\urllib3-1.10.4-py3.4.egg\urllib3\util\retry.py", line 271, in increment
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='fbcdn-profile-a.akamaihd.net', port=443): Max retries exceeded with url: /hprofile-ak-xfp1/v/t1.0-1/p32x32/1795556_454025928112183_7252557420120250438_n.jpg?oh=2f63a1f58e29fd34605626a87fdd7202&oe=55EA9D3F&__gda__=1445088810_e91fee57938ed896996ea686b6a0e628 (Caused by ReadTimeoutError("HTTPSConnectionPool(host='fbcdn-profile-a.akamaihd.net', port=443): Read timed out. (read timeout=5)",))
```

**Comment by Lukasa** [comment]
Ok cool, so we're getting a read timeout. This means the TCP connection is getting established but the other side isn't sending any data. There could be a few reasons for this happening. Are you behind a proxy?

**Comment by cattleyavns** [comment]
>> 

Ok cool, so we're getting a read timeout. This means the TCP connection is getting established but the other side isn't sending any data. There could be a few reasons for this happening. Are you behind a proxy?

I'm tested the code above on two computers, without any proxy server and both computer show the same result.

**Comment by Lukasa** [comment]
urllib3 doesn't support SPDY, but it's clearly orthogonal to the problem because I'm having no difficulty on my end.

Are you familiar with tcpdump?

**Comment by cattleyavns** [comment]
No, I'm not, but if I want to analyze TCP traffic, I will use sniffer tool from Nirsoft.

**Comment by Lukasa** [comment]
What I want to understand is how the TLS handshake is going. It seems to me like the remote end is falling over, and I'm running very recent versions of OpenSSL that you may not be running.

**Comment by Lukasa** [comment]
While we're here, can you tell me the output of `import ssl; ssl.OPENSSL_VERSION`?

**Comment by t-8ch** [comment]
Could you check if IPv6 is in play? I had issues with it and some of facebooks CDN servers

**Comment by Lukasa** [comment]
This would be the weirdest IPv6 bug, to successfully handshake but fail to transmit/receive.

**Comment by cattleyavns** [comment]
Okay, can you explain what I should do with tcpdump, I don't even know where to download the software, it is Linux or Windows dependent ? I'm using Windows.

Thanks, I don't think OpenSSL causing this problem, urllib3 on Python 3 don't really need OpenSSL, but ssl library is enough as far as I know.

**Comment by Lukasa** [comment]
When I say OpenSSL I literally mean the C library, not the Python wrapper called PyOpenSSL. That's why I asked you to print the version from the `ssl` library.

If you're on Windows you can actually use the Wireshark tool more easily than tcpdump. You'll need to capture the request and filter the traffic to be only that, then save the pcap output. =)

**Comment by cattleyavns** [comment]
Okay, I will try! Sorry I mistakenly deleted my previous post.

**Comment by cattleyavns** [comment]
Here is my ssl.OPENSSL version:
```
>>> import ssl
>>> print(ssl.OPENSSL_VERSION)
OpenSSL 1.0.1l 15 Jan 2015
```

**Comment by cattleyavns** [comment]
My PcapNG file from Wireshark, download and rename file extension to 1.pcapng:

![1 - copy](https://cloud.githubusercontent.com/assets/10792823/8369951/3e917b92-1bfc-11e5-8581-820d334a5d02.png)

**Comment by Lukasa** [comment]
Well that's thoroughly perplexing.

The application sends 1600 bytes of encrypted data is its response. For both you and I that gets split over two TCP segments (because it's larger than my 1514 byte MTU and your 1466 byte MTU). For me I receive those two TCP segments basically right at the same time, less than 1ms apart. For you, you seem to receive the *first* segment, but the second doesn't arrive until well after the connection is torn down (you tear the connection down at ~15 seconds in, the missing TCP segment arrives at 15.7 seconds).

It looks like this TCP segment is getting stuck in some of your network equipment. It can't be anything else, because you and the remote end have already done a nice little FIN dance to do clean shutdown.

This is pretty consistent: each other test run has the same behaviour (issue the request, get the first response segment immediately, tear down after 5 seconds, receive the missing TCP segment later. Twice it's 0.7 seconds later but at least once it's nearly 3 seconds later).

I'd be interested to see the connection from Firefox, because it seems like your network is *totally* screwing this up. There's a packet getting stuck *somewhere* between you and Akamai. When you tested with/without a proxy, were you on the same network? Are you working on a laptop? If so, can you try a different network connection and try again?

**Comment by cattleyavns** [comment]
>> I'd be interested to see the connection from Firefox, because it seems like your network is totally screwing this up. There's a packet getting stuck somewhere between you and Akamai. When you tested with/without a proxy, were you on the same network? Are you working on a laptop? If so, can you try a different network connection and try again?

Okay, Here is my Firefox pcapng:

I'm sure with Firefox, no proxy it works just above than fine, I already clear my cache or Ctrl + F5 to make sure I redownload the URL.

Pcapng file, please rename it to "2.pcapng": 
![2 - copy](https://cloud.githubusercontent.com/assets/10792823/8373864/77bc7654-1c2b-11e5-9e7b-c09c85ae8427.png)

Image: http://i.imgur.com/fwsBoUk.png

**Comment by cattleyavns** [comment]
>> I'd be interested to see the connection from Firefox, because it seems like your network is totally screwing this up. There's a packet getting stuck somewhere between you and Akamai. When you tested with/without a proxy, were you on the same network? Are you working on a laptop? If so, can you try a different network connection and try again?

Okay, Here is my Firefox pcapng:

I'm sure with Firefox, no proxy it works just above than fine, I already clear my cache or Ctrl + F5 to make sure I redownload the URL.

Pcapng file, please rename it to "2.pcapng": 
![22 - copy](https://cloud.githubusercontent.com/assets/10792823/8373918/efbbd8a2-1c2b-11e5-9fb2-a1b0d6dfcbdc.png)


Image: http://i.imgur.com/gMPt4k5.png

Edit: Repost

**Comment by Lukasa** [comment]
Yup, so that Firefox behaviour is the same as mine. I really suspect there's a middlebox trying to do DPI and screwing it up. Can you answer the other questions in my post above? =)

**Comment by cattleyavns** [comment]
>> Can you answer the other questions in my post above? =)

But Firefox can load that URL fine, but Python cannot with urllib3 or urllib.request so I think other network might be the same.
I will try to get another network and test, I will post my result as soon as I can.

**Comment by Lukasa** [comment]
I doubt another network will be the same.

The key difference, more than anything else, is that Firefox does not use the same network stack as the various Python programs you've tested. In particular, it uses NSS, not OpenSSL. This means it builds TLS fundamentally differently. This will lead to a different TCP segment profile, and may therefore be acceptable to whatever middlebox appears to be ruining this flow.

**Comment by Lukasa** [comment]
Bump. =)

**Comment by cattleyavns** [comment]
After that long time, I still have the same problem.

I got another network (different IP, diffirent ISP) but at this time it is bugging, so I will try to fix it and report, maybe next day.

**Comment by lahwaacz** [comment]
In [this comment](https://github.com/shazow/urllib3/issues/656#issuecomment-115484197), there is first `TypeError: getresponse() got an unexpected keyword argument 'buffering'` and only after that the timeout: `socket.timeout: The read operation timed out`
Is this normal or another problem? I don't see how the two exceptions might be related...

**Comment by Lukasa** [comment]
We can't do that in Python 3 because it's invalid syntax in Python 2.

**Comment by saper** [comment]
(5 years later...) What wonders me is why suddenly your router at 192.168.1.1 (MAC fc:b6:98:60:f3:50) starts issuing ARP query for your machine as if it forgot it was communicating with you all the time. Maybe it lost the missing segment?

**Comment by lahwaacz** [comment]
> We can't do that in Python 3 because it's invalid syntax in Python 2.

Since Python 2 is finally EOL, you can do that.

**Comment by sigmavirus24** [comment]
> Since Python 2 is finally EOL, you can do that.

@lahwaacz actually urllib3 is continuing to support Python 2 beyond its EOL so, no... it can't

**Comment by sethmlarson** [comment]
Going to close this issue as not reproducible and likely not a urllib3 issue. Let me know if there's more information.
